File: book-library/src/main/java/com/accenture/booklibrary/service/BookService.java
package com.accenture.booklibrary.service;

import com.accenture.booklibrary.model.Book;
import com.accenture.booklibrary.repository.BookRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class BookService {

    private final BookRepository bookRepository;

    public BookService(BookRepository bookRepository) {
        this.bookRepository = bookRepository;
    }

    @Transactional(readOnly = true)
    public Book getBookByTitle(String title) {
        return bookRepository.findByTitle(title)
                .orElseThrow(() -> new NoSuchElementException("There is not book with title: %s".formatted(title)));
    }

    @Transactional(readOnly = true)
    public List<Book> getBooksByAuthor(String author) {
        return bookRepository.findByAuthor(author);
    }

    @Transactional
    public String createBook(Book book) {
        Book b = bookRepository.save(book);
        return "/api/v1/books/" + b.getTitle();
    }

    @Transactional
    public void deleteBook(Long id) {
        bookRepository.deleteById(id);
    }

    @Transactional(readOnly = true)
    public Set<String> getCategories() {
        return bookRepository.findAll()
                .stream()
                .map(Book::getCategory)
                .collect(Collectors.toSet());
    }

    @Transactional(readOnly = true)
    public Set<String> getAuthors() {
        return bookRepository.findAll()
                .stream()
                .map(Book::getAuthor)
                .collect(Collectors.toSet());
    }

    @Transactional(readOnly = true)
    public List<String> getTitles() {
        return bookRepository.findAll()
                .stream()
                .map(Book::getTitle)
                .toList();
    }
}

File: book-library-frontend/src/app/services/book.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import {Book} from '../model/book.model';
import {BookRequest} from '../model/book.request.model';

@Injectable({
  providedIn: 'root'
})
export class BooksService {
  private readonly apiUrl = 'http://localhost:8080/api/v1/books';

  constructor(private readonly http: HttpClient) {}

  getTitles(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/titles`);
  }

  getCategories(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/categories`);
  }

  getAuthors(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/authors`);
  }

  getBookByTitle(title: string): Observable<Book> {
    return this.http.get<Book>(`${this.apiUrl}/title/${title}`);
  }

  getBooksByAuthor(author: string): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.apiUrl}/author/${author}`);
  }

  createBook(book: BookRequest): Observable<BookRequest> {
    return this.http.post<BookRequest>(`${this.apiUrl}`, book);
  }

  deleteBook(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
